File: frontend/lib/features/goals/domain/monthly_goal.dart
class MonthlyGoal {
  const MonthlyGoal({
    required this.id,
    required this.userId,
    required this.month,
    required this.title,
    required this.target,
    required this.unit,
    required this.progress,
  });

  final String id;
  final String userId;

  /// First day of the goal's month, e.g. 2026-06-01. Time-of-day is ignored.
  final DateTime month;

  final String title;

  /// Fractional so a task can earn partial credit (e.g. 0.25 of an
  /// "hours" goal from a 15-minute task).
  final num target;
  final String unit;
  final num progress;

  bool get isComplete => progress >= target;

  static DateTime monthKeyFor(DateTime date) => DateTime(date.year, date.month, 1);

  static const maxActiveGoals = 5;
}

extension FormattedNum on num {
  /// Formats a whole number without a trailing ".0" (e.g. 5.0 -> "5"),
  /// while still showing fractional amounts plainly (e.g. 1.25 -> "1.25").
  String get formatted => this % 1 == 0 ? toInt().toString() : toString();
}

File: backend/app/domain/monthly_goal.py
from dataclasses import dataclass
from datetime import date
from uuid import UUID

MAX_ACTIVE_GOALS = 5


@dataclass
class MonthlyGoal:
    id: UUID
    user_id: UUID
    month: date  # always the 1st of the month; time-of-day is irrelevant
    title: str
    target: int
    unit: str
    progress: int = 0

    @property
    def is_complete(self) -> bool:
        return self.progress >= self.target


def month_key_for(value: date) -> date:
    return value.replace(day=1)
